pandas/core/arrays/numpy_.py

```diff
@@ -507,6 +507,46 @@ def to_numpy(
 
         return result
 
+    def _validate_setitem_value(self, value):
+        if type(value) == int:
+            if (
+                self.dtype == NumpyEADtype("int64")
+                or self.dtype == NumpyEADtype("float64")
+                or self.dtype == NumpyEADtype("uint16")
+                or self.dtype == NumpyEADtype("object")
+                or self.dtype is None
+            ):
+                return value
+        elif type(value) == float:
+            if (
+                self.dtype == NumpyEADtype("float64")
+                or self.dtype == NumpyEADtype("object")
+                or self.dtype is None
+            ):
+                return value
+        elif type(value) == str:
+            if (
+                self.dtype == NumpyEADtype("str")
+                or self.dtype == NumpyEADtype("object")
+                or self.dtype is None
+                or self.dtype == NumpyEADtype("U32")
+            ):
+                return value
+        elif NumpyEADtype(type(value)) == NumpyEADtype(self.dtype) or NumpyEADtype(
+            type(value)
+        ) == NumpyEADtype(type(value)):
+            return value
+        else:
+            if self.dtype == NumpyEADtype("object") or self.dtype is None:
+                return value
+        if NumpyEADtype(type(value)) != self.dtype:
```

Comment:
I think you can just add a check for `NumpyEADtype(type(value)) == self.dtype` as an `or` when you do `NumpyEADtype(type(value)) == NumpyEADtype(self.dtype)`

Reply:
Then you can just raise a TypeError without any more checks